Docs about AWS hubs and CI/CD automation (#717)
2i2c-org/infrastructure: docs/howto/operate/setup-ci-cd-aws.md

(operate:setup-ci-cd-aws)=

> # Set up the CI/CD system in an AWS cluster

Reviewer: Since we might want to set up CI/CD auto-deploys across all of the cloud providers, perhaps we should create a top-level section under `Topics` called `Automatic deployment with CI/CD`, and then underneath that are pages for each of the major cloud providers. (this assumes that it will be slightly different for AWS, GCP, and Azure).

Does that make sense to you?

> The pilot-hubs deployer code supports [a new `auth_aws` function](https://github.com/2i2c-org/pilot-hubs/blob/e96e7bcded187870dc2e07d6626de8a12586ed32/deployer/hub.py#L126)
> to get a temporary kubeconfig file that allows a successful connection with the
> kubernetes cluster.

Reviewer 2: ```suggestion
The pilot-hubs deployer code supports [a new `auth_aws` function](https://github.com/2i2c-org/pilot-hubs/blob/e96e7bcded187870dc2e07d6626de8a12586ed32/deployer/hub.py#L126)
that can export a temporary kubeconfig file to allows a successful connection with the
kubernetes cluster.
```

> It is actually expecting a 2i2c generated IAM entity (the `deployer` user) and its
> AWS credentials (that we provide in a sops encrypted json file) to properly set them up
> as the AWS `Access Key ID` and `Secret Access Key` as environment variables and
> successfully perform operations in the cluster as the `deployer` user.

Reviewer 2: Maybe use shorter phrases?
```suggestion
The function expects a 2i2c generated IAM entity (a `deployer` user) and its AWS credentials (provided as a sops encrypted `json` file). These will be used to properly set up the AWS `Access Key ID` and `Secret Access Key` environment variables and to allow performing operations in the cluster as the `deployer` user.
```

There are some requisites for being able to successfully and automatically deploy hubs
on AWS clusters using the provided functionality:

> 1. Find out if you are working with a [AWS EKS-based](https://aws.amazon.com/eks/) or a

Reviewer: Given that there are no other sections on this page, perhaps the numbers could be headers (`##`) instead of numbered lists, and then a reader could click through them more quickly?

[kops-based](https://kops.sigs.k8s.io/getting_started/aws/) cluster
  
   We are currently deploying `EKS`-based (Carbonplan) and `kops`-based (Farallon,
   Openscapes) clusters. Whether to create one or another depends on the cluster
   operational needs and some other factors outlined and being discussed in a dedicated
   [pilot-hub issue](https://github.com/2i2c-org/pilot-hubs/issues/431).
   You need to know and understand which sort of cluster are you working with to properly
   set up the config file in step 4.

   Look at [the AWS cluster setup instructions](new-cluster:aws) if you are interested in
   setting up a `kops`-based cluster.

   :::{note}
   Documentation about creating an `EKS`-based cluster has not yet being written
   :::

> 2. Create a `deployer` user with enough permissions for being able to automatically

Reviewer: Could this link to an AWS doc about what users are etc if people want to learn more?

     clusterType: eks
     clusterName: carbonplanhub
     region: us-west-2
   ```

   You may notice you need to provide `aws` as the provider and then, under the `aws` key,
   there is some additional keys to be filled:

     * `key`, that one points to the encrypted sops files containing the `deployer` credentials
     * `clusterType`, whether it is an `eks` or `kops`-based cluster
     * `clusterName`, the name of the kubernetes cluster
     * `region`, the AWS region where the cluster lives

   The `kops`-based cluster need some additional information containing the state of the cluster

   This is short snippet from the Farallon config file

   ```yaml
   provider: aws
   aws:
     key: secrets/farallon.json
     clusterType: kops
     clusterName: farallonhub.k8s.local
     region: us-east-2
     stateStore: s3://2i2c-farallon-kops-state
   ```

   Notice the additional `stateStore` key pointing to the kops state living on a specific s3 bucket.

5. Check if the cluster was added to the [CI deploy-hubs workflow](https://github.com/2i2c-org/pilot-hubs/blob/e96e7bcded187870dc2e07d6626de8a12586ed32/.github/workflows/deploy-hubs.yaml#L31-L36)

   :::{note}

>    We are conditionally installing `kops` if the provider is `aws` even with EKS-based cluster.

Reviewer 2: Maybe a short sentence describing why.
